### DISCUSSION TITLE
[Question] [Win10] How can I draw silently and export the rendered image into opengl texture format?

### DISCUSSION THREAD

**Discussion by Rullec** [alandefreitas/matplotplusplus#199] [Q&A]
Hi guys,
   Thanks for your excellent work on this lib, I love it a lot. I work on win10 and gnuplot backend, and really eager to know:
   1. how can `matplotplusplus` draw silently? at this moment it pops up a window, which is not my intention.
   2. how can I export the rendered image into OpenGL texture format (an array of float numbers, given width and height), instead of saving them to "*png" or "*jpg" at disk.

Thanks a lot!
Bests,
Xudong

**Comment by DaMilyutin** [comment]
I have same question.
Maybe you found the answer elsewhere? Can you share it, please?

**Reply by Rullec** [reply]
Hi Daniel,
   Instead of `matplotplusplus`, I then switch to another library called `ImPlot`, which can do plotting works over my screen. Then I got the rendered image by retrieving framebuffer from OpenGL.

**Reply by Rullec** [reply]
I am glad to offer my help if you are insterested in this process. ;-)

**Reply by DaMilyutin** [reply]
Thank you for the answer and for the help offer.
I'm looking for some more fancy-looking plots and GUI in my pet project.
I used Qt QCustomPlot some time ago. But...
I want it to be lightweight (no big dll dependencies, smaller exe) and use free libraries and frameworks.
I was in search for libraries and tools for that. 

For GUI I've discovered sixtyfps. Quite intresting alternative to Qt/QML which binds both to C+++ and Rust.
But I had some difficulties with passing data around between C++ and .60 (their markup lang which is compiled from Rust).
I need to pass arrays of doubles.

What struck me lately is to write prototype totally in Rust. And thus it would be easier to manage.

BTW, there is Rust library named plotters: https://github.com/38/plotters
If you want more fancy looking plots and 3d plots you may either use it from Rust or write some C/C++ bindings for it.
It can output to buffer, too. Ex, see: https://github.com/sixtyfpsui/sixtyfps/tree/master/examples/plotter

**Reply by Rullec** [reply]
Thanks for your kine reply!

**Reply by Rullec** [reply]
Though I have no idea about `Rust`, but it indeed looks fancinating! I will certainly try `sixtyfps` when it is needed in the future. Good luck!

**Reply by HarrisonLeece** [reply]
@Rullec I want to do something similar.  I'm in such need for it that I am learning OpenGL.  So far I found this link, https://lencerf.github.io/post/2019-09-21-save-the-opengl-rendering-to-image-file/  which may be helpful, if you are able to use OpenGL with some proficiency already.  Unfortunately, I've not yet written a program that saves the image yet.

**Comment by alandefreitas** [comment]
Hi @Rullec 

matplot++ doesn't work at this level of abstraction. Backends could but are not required to allow exposing the raster image buffer as binary data. In fact, they are not even required to be raster images, as they could be vector images.  

If you need plots inside an OpenGL application for debugging, it'd easier to use something at this level of abstraction, such ImGui + ImPlot. If it's not for debugging, then you'd have to draw the plot yourself depending on the level of customization you need. If you need plots in a Qt application, it'd be easier to use Qt Charts. And so on...

**Reply by Rullec** [reply]
Thanks for your reply!

**Reply by Rullec** [reply]
Yes, I think you can totally correct. The design methodology of `matplot++` is not quite suitable for my case. 
Still thanks you work on this project ;-)
